File: Fun Games/colonel_blotto/colonel_blotto.cpp
#include <bits/stdc++.h>
using namespace std;

vector<int> allocateSoldiers(int totalSoldiers, int totalTowers, int player) {
    vector<int> allocations;
    int remainingSoldiers = totalSoldiers;

    for (int i = 0; i < totalTowers - 1; ++i) {
        int maxAllocation = min(remainingSoldiers, totalSoldiers - (totalTowers - i - 1));
        int allocation;

        cout << "Player " << player << ", enter the number of soldiers for Tower " << i + 1 << " (1-" << maxAllocation << "): ";
        cin >> allocation;

        while (allocation < 1 || allocation > maxAllocation) {
            cout << "Invalid input. Please enter a number between 1 and " << maxAllocation << ": ";
            cin >> allocation;
        }

        allocations.push_back(allocation);
        remainingSoldiers -= allocation;
    }

    allocations.push_back(remainingSoldiers);
    return allocations;
}

int compareTowerConfigurations(const vector<int>& config1, const vector<int>& config2, int totalTowers) {
    int count = 0;

    for (int i = 0; i < totalTowers; ++i) {
        if (config1[i] > config2[i]) {
            count++;
        }
    }

    return count;
}

void playGame() {
    int totalSoldiers, totalTowers, totalPlayers;

    cout << "Enter the total number of soldiers: ";
    cin >> totalSoldiers;

    cout << "Enter the total number of towers: ";
    cin >> totalTowers;

    cout << "Enter the number of players: ";
    cin >> totalPlayers;

    // Store tower configurations for each player
    vector<vector<int>> playerAllocationsList;

    for (int player = 0; player < totalPlayers; ++player) {
        cout << "\nPlayer " << player + 1 << "'s turn:\n";
        vector<int> playerAllocations = allocateSoldiers(totalSoldiers, totalTowers, player + 1);
        playerAllocationsList.push_back(playerAllocations);
    }

    // Count wins for each player
    vector<int> playerWins(totalPlayers, 0);

    for (int player1 = 0; player1 < totalPlayers; ++player1) {
        for (int player2 = player1 + 1; player2 < totalPlayers; ++player2) {
            // Compare tower configurations and determine the winner of the pairwise comparison
            int comparisonResult = compareTowerConfigurations(playerAllocationsList[player1], playerAllocationsList[player2], totalTowers);

            if (comparisonResult > totalTowers / 2) {
                playerWins[player1]++;
            } else {
                playerWins[player2]++;
            }
        }
    }

    cout << "\nRound Results:\n";
    for (int player = 0; player < totalPlayers; ++player) {
        cout << "Player " << player + 1 << " wins " << playerWins[player] << " pairwise comparisons.\n";
    }

    // Find the round winner
    int roundWinner = max_element(playerWins.begin(), playerWins.end()) - playerWins.begin() + 1;
    cout << "\nPlayer " << roundWinner << " is the winner of the round!\n";
}

int main() {
    playGame();
    return 0;
}

File: Fun Games/colonel_blotto/colonel_blotto.py
import itertools

def allocate_soldiers(total_soldiers, total_towers, player):
    allocations = []
    remaining_soldiers = total_soldiers

    for i in range(total_towers - 1):
        max_allocation = min(remaining_soldiers, total_soldiers - (total_towers - i - 1))
        allocation = int(input(f"Player {player}, enter the number of soldiers for Tower {i + 1} (1-{max_allocation}): "))

        while allocation < 1 or allocation > max_allocation:
            print(f"Invalid input. Please enter a number between 1 and {max_allocation}.")
            allocation = int(input(f"Player {player}, enter the number of soldiers for Tower {i + 1} (1-{max_allocation}): "))

        allocations.append(allocation)
        remaining_soldiers -= allocation

    allocations.append(remaining_soldiers)
    return allocations

def play_game():
    total_soldiers = int(input("Enter the total number of soldiers: "))
    total_towers = int(input("Enter the total number of towers: "))
    total_players = int(input("Enter the number of players: "))

    # Store tower configurations for each player
    player_allocations_list = []

    for player in range(total_players):
        print(f"\nPlayer {player + 1}'s turn:")
        player_allocations = allocate_soldiers(total_soldiers, total_towers, player + 1)
        player_allocations_list.append(player_allocations)

    # Generate all pairwise combinations of players
    player_combinations = list(itertools.combinations(range(total_players), 2))

    # Count wins for each player
    player_wins = [0] * total_players

    for combination in player_combinations:
        player1, player2 = combination
        player1_allocations = player_allocations_list[player1]
        player2_allocations = player_allocations_list[player2]

        # Compare tower configurations and determine the winner of the pairwise comparison
        comparison_result = sum(1 for p1, p2 in zip(player1_allocations, player2_allocations) if p1 > p2)

        if comparison_result > total_towers // 2:
            player_wins[player1] += 1
        else:
            player_wins[player2] += 1

    print("\nRound Results:")
    for player in range(total_players):
        print(f"Player {player + 1} wins {player_wins[player]} pairwise comparisons.")

    round_winner = player_wins.index(max(player_wins)) + 1
    print(f"\nPlayer {round_winner} is the winner of the round!")

if __name__ == "__main__":
    play_game()
